stan/math/prim/scal/meta/is_container.hpp

```diff
@@ -0,0 +1,21 @@
+#ifndef STAN_MATH_PRIM_SCAL_META_IS_CONTAINER_HPP
+#define STAN_MATH_PRIM_SCAL_META_IS_CONTAINER_HPP
+
+#include <stan/math/prim/scal/meta/bool_constant.hpp>
+#include <stan/math/prim/scal/meta/disjunction.hpp>
+#include <stan/math/prim/scal/meta/is_eigen.hpp>
+#include <stan/math/prim/scal/meta/is_vector.hpp>
+#include <type_traits>
+
+namespace stan {
+
+/**
+ * Deduces whether type is convertible to string
+ * @tparam T type to check
+ */
+template <typename Container>
+using is_container = bool_constant<
+    math::disjunction<is_eigen<Container>, is_std_vector<Container>>::value>;
```

Comment:
`is_vector_like` would fail for Eigen matrices right? I have to look at that. If so then maybe we should change the name of `is_vector_like` to `is_container`? That seems to be more of what it actually looks for (or at least guarantees is true)

Reply:
It actually returns true for almost any Eigen type, the ```is_eigen``` type trait tests whether the type is derived from ```EigenBase```, ```DenseBase```, and ```MatrixBase``` - which would cover just about everything, I think

Reply:
Ended up backporting `void_t` and `is_detected` from C++17 so that `is_vector_like` checks for `operator[int]` now (which I think is a better definition. Still need to write tests for those tho' so I'll ping you for review